File: examples/joern-test-data/javascript/sample.js
/**
 * Sample JavaScript code for Joern CPG analysis testing
 * Demonstrates: classes, inheritance, async/await, closures, modules
 */

// Shape base class
class Shape {
  #color;

  constructor(color = 'white') {
    this.#color = color;
  }

  get color() {
    return this.#color;
  }

  set color(value) {
    this.#color = value;
  }

  get name() {
    throw new Error('Subclasses must implement name');
  }

  area() {
    throw new Error('Subclasses must implement area()');
  }

  perimeter() {
    throw new Error('Subclasses must implement perimeter()');
  }

  printInfo() {
    return `${this.name}: area=${this.area().toFixed(2)}, perimeter=${this.perimeter().toFixed(2)}`;
  }
}

// Rectangle class
class Rectangle extends Shape {
  #width;
  #height;

  constructor(width, height, color = 'white') {
    super(color);
    this.#width = width;
    this.#height = height;
  }

  get name() {
    return 'Rectangle';
  }

  get width() {
    return this.#width;
  }

  get height() {
    return this.#height;
  }

  area() {
    return this.#width * this.#height;
  }

  perimeter() {
    return 2 * (this.#width + this.#height);
  }

  isSquare() {
    return Math.abs(this.#width - this.#height) < 0.0001;
  }
}

// Circle class
class Circle extends Shape {
  static PI = 3.14159265358979;

  #radius;

  constructor(radius, color = 'white') {
    super(color);
    this.#radius = radius;
  }

  get name() {
    return 'Circle';
  }

  get radius() {
    return this.#radius;
  }

  get diameter() {
    return 2 * this.#radius;
  }

  area() {
    return Circle.PI * this.#radius ** 2;
  }

  perimeter() {
    return 2 * Circle.PI * this.#radius;
  }
}

// ShapeManager class with functional features
class ShapeManager {
  #shapes = [];

  addShape(shape) {
    this.#shapes.push(shape);
    return this; // Enable chaining
  }

  totalArea() {
    return this.#shapes.reduce((sum, shape) => sum + shape.area(), 0);
  }

  totalPerimeter() {
    return this.#shapes.reduce((sum, shape) => sum + shape.perimeter(), 0);
  }

  filter(predicate) {
    return this.#shapes.filter(predicate);
  }

  findByColor(color) {
    return this.filter((shape) => shape.color === color);
  }

  forEach(callback) {
    this.#shapes.forEach(callback);
  }

  map(callback) {
    return this.#shapes.map(callback);
  }

  printAll() {
    this.forEach((shape) => console.log(shape.printInfo()));
  }

  // Async processing
  async processAsync() {
    const delay = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

    const results = await Promise.all(
      this.#shapes.map(async (shape) => {
        await delay(10); // Simulate async work
        return `Processed: ${shape.printInfo()}`;
      })
    );

    return results;
  }
}

// Factory function with closure
function createShapeFactory() {
  let createdCount = 0;

  return {
    createRectangle(width, height, color) {
      if (width <= 0 || height <= 0) {
        throw new Error('Dimensions must be positive');
      }
      createdCount++;
      return new Rectangle(width, height, color);
    },

    createCircle(radius, color) {
      if (radius <= 0) {
        throw new Error('Radius must be positive');
      }
      createdCount++;
      return new Circle(radius, color);
    },

    getCreatedCount() {
      return createdCount;
    },
  };
}

// Higher-order function for filtering
const createAreaFilter = (minArea) => (shape) => shape.area() > minArea;

// Main execution
async function main() {
  const manager = new ShapeManager();
  const factory = createShapeFactory();

  manager
    .addShape(factory.createRectangle(5.0, 3.0, 'red'))
    .addShape(factory.createCircle(2.5, 'blue'))
    .addShape(factory.createRectangle(4.0, 4.0, 'red'));

  console.log('All shapes:');
  manager.printAll();

  console.log(`\nTotal area: ${manager.totalArea().toFixed(2)}`);
  console.log(`Total perimeter: ${manager.totalPerimeter().toFixed(2)}`);

  // Using higher-order function
  const largeShapes = manager.filter(createAreaFilter(15));
  console.log('\nLarge shapes (area > 15):');
  largeShapes.forEach((shape) => console.log(`  ${shape.printInfo()}`));

  // Find by color
  const redShapes = manager.findByColor('red');
  console.log('\nRed shapes:');
  redShapes.forEach((shape) => console.log(`  ${shape.printInfo()}`));

  // Async processing
  console.log('\nAsync processing:');
  const results = await manager.processAsync();
  results.forEach((result) => console.log(`  ${result}`));

  console.log(`\nFactory created ${factory.getCreatedCount()} shapes`);
}

main().catch(console.error);

// Export for module usage
module.exports = { Shape, Rectangle, Circle, ShapeManager, createShapeFactory };

File: examples/joern-test-data/ruby/sample.rb
# Sample Ruby code for Joern CPG analysis testing
# Demonstrates: classes, modules, mixins, blocks, lambdas

# Shape module with shared functionality
module Printable
  def print_info
    "#{name}: area=#{area.round(2)}, perimeter=#{perimeter.round(2)}"
  end
end

# Abstract-like base class
class Shape
  include Printable

  attr_accessor :color

  def initialize(color = 'white')
    @color = color
  end

  def area
    raise NotImplementedError, 'Subclasses must implement area'
  end

  def perimeter
    raise NotImplementedError, 'Subclasses must implement perimeter'
  end

  def name
    self.class.name
  end
end

# Rectangle class
class Rectangle < Shape
  attr_reader :width, :height

  def initialize(width, height, color = 'white')
    super(color)
    @width = width
    @height = height
  end

  def area
    @width * @height
  end

  def perimeter
    2 * (@width + @height)
  end

  def square?
    @width == @height
  end
end

# Circle class
class Circle < Shape
  PI = 3.14159265358979

  attr_reader :radius

  def initialize(radius, color = 'white')
    super(color)
    @radius = radius
  end

  def area
    PI * @radius ** 2
  end

  def perimeter
    2 * PI * @radius
  end

  def diameter
    2 * @radius
  end
end

# ShapeManager with functional programming features
class ShapeManager
  def initialize
    @shapes = []
  end

  def add_shape(shape)
    @shapes << shape
    self
  end

  def total_area
    @shapes.reduce(0.0) { |sum, shape| sum + shape.area }
  end

  def total_perimeter
    @shapes.sum(&:perimeter)
  end

  def each(&block)
    @shapes.each(&block)
  end

  def map(&block)
    @shapes.map(&block)
  end

  def select(&block)
    @shapes.select(&block)
  end

  def find_by_color(color)
    select { |shape| shape.color == color }
  end

  def print_all
    each { |shape| puts shape.print_info }
  end

  # Lambda for custom filtering
  def filter_by(&predicate)
    @shapes.select(&predicate)
  end
end

# Main execution
if __FILE__ == $PROGRAM_NAME
  manager = ShapeManager.new

  manager
    .add_shape(Rectangle.new(5.0, 3.0, 'red'))
    .add_shape(Circle.new(2.5, 'blue'))
    .add_shape(Rectangle.new(4.0, 4.0, 'red'))

  puts 'All shapes:'
  manager.print_all

  puts "\nTotal area: #{manager.total_area.round(2)}"
  puts "Total perimeter: #{manager.total_perimeter.round(2)}"

  # Using lambda for filtering
  large_shapes = manager.filter_by { |s| s.area > 15 }
  puts "\nLarge shapes (area > 15):"
  large_shapes.each { |s| puts "  #{s.print_info}" }

  # Find red shapes
  red_shapes = manager.find_by_color('red')
  puts "\nRed shapes:"
  red_shapes.each { |s| puts "  #{s.print_info}" }
end
